refactor: Use C++17 std::clamp

hebasto opened the pull request:
This PR improves code readability.

Only safe cases with guarantee of no [undefined behavior](https://en.cppreference.com/w/cpp/algorithm/clamp).

hebasto committed:
refactor: Use C++17 std::clamp

This change improves code readability.

theStack commented:
Concept ACK

jarolrod commented:
Concept ACK

`std::clamp` is standard in C++17 and is certainly a nice cleanup versus the existing code. Additionally, this is not a breaking change for windows builders as this is available since VS2015.

kiminuo commented:
Concept ACK

I wonder whether this is a good candidate or not: https://github.com/bitcoin/bitcoin/blob/58c25bdcea9b95348e639180932b388d4bda8157/src/rpc/blockchain.cpp#L1859

hebasto commented:
@kiminuo 

> I wonder whether this is a good candidate or not:
> 
> https://github.com/bitcoin/bitcoin/blob/58c25bdcea9b95348e639180932b388d4bda8157/src/rpc/blockchain.cpp#L1859

Some tests fail with this line being changed.

kiminuo commented:
> @kiminuo
> 
> > I wonder whether this is a good candidate or not:
> > https://github.com/bitcoin/bitcoin/blob/58c25bdcea9b95348e639180932b388d4bda8157/src/rpc/blockchain.cpp#L1859

In case you think it is a good candidate for changing to `std::clamp`, it would be helpful to post how you changed the line and a test failure details.

hebasto commented:
@kiminuo 
> In case you think it is a good candidate for changing to `std::clamp`, it would be helpful to post how you changed the line and a test failure details.

Here are the [branch](https://github.com/hebasto/bitcoin/commits/210925-clamp-DEMO) and [tests](https://cirrus-ci.com/build/4764240100720640).

laanwj commented:
Ouch-if tests fail that means the behavior is not the same. I think we should be careful here. For the C++11 transition we even had an explicit rule not to do all-over-the-place code modernization like this, it's very easy to accidentally introduce a bug. I'm all for using `clamp` but might be better to introduce it when code is changed anyway.

MarcoFalke commented:
This is not a trivial replacement. With std::min/max the argument order doesn't matter (min(a,b)==min(b,a)). However, with `std::clamp`, the order must be exactly right and there is only one correct order.

martinus commented:
I think the test has failed because `std::clamp`'s behavior is undefined when when the lo value is bigger than the hi value. For `std::clamp(blockcount, 0, pindex->nHeight - 1)` this is possible when `pindex->nHeight==0`

martinus commented on src/checkqueue.h:
```diff
@@ -109,7 +109,7 @@ class CCheckQueue
                 // * Try to account for idle jobs which will instantly start helping.
                 // * Don't do batches smaller than 1 (duh), or larger than nBatchSize.
-                nNow = std::max(1U, std::min(nBatchSize, (unsigned int)queue.size() / (nTotal + nIdle + 1)));
+                nNow = std::clamp<unsigned int>(queue.size() / (nTotal + nIdle + 1), 1, nBatchSize);
```
Not sure if this can ever happen, but when `nBatchSize==0` this will now have undefined behavior. Previously the result used to be 1.

hebasto replied:
Right. Would `assert(nBatchSize >= 1);` in the `CCheckQueue` constructor be enough?

martinus replied:
I'd put it right above the clamp

hebasto replied:
I thought again, and put the assertion into the constructor because it is constructor's job to establish safe invariants for a class instance.

martinus commented on src/init.cpp:
```diff
@@ -880,7 +881,7 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     int fd_max = FD_SETSIZE;
 #endif
-    nMaxConnections = std::max(std::min<int>(nMaxConnections, fd_max - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE), 0);
+    nMaxConnections = std::clamp<int>(nMaxConnections, 0, fd_max - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE);
```
If `fd_max - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE` is negative then `nMaxConnections` used to be 0, but now the result is undefined

hebasto replied:
This change [dropped](https://github.com/bitcoin/bitcoin/pull/23095#issuecomment-928946761).

hebasto commented:
Updated 82c4e71025d0958b4066325731874863e1323a0e -> bf4838e6456fa744911a886356529656a16c19ed ([pr23095.04](https://github.com/hebasto/bitcoin/commits/pr23095.04) -> [pr23095.05](https://github.com/hebasto/bitcoin/commits/pr23095.05)):

- rebased
- addressed comments

martinus reviewed (approved):
Code review ACK bf4838e6456fa744911a886356529656a16c19ed

laanwj commented on src/miner.cpp:
```diff
@@ -64,7 +64,7 @@ BlockAssembler::BlockAssembler(CChainState& chainstate, const CTxMemPool& mempoo
     blockMinFeeRate = options.blockMinFeeRate;
     // Limit weight to between 4K and MAX_BLOCK_WEIGHT-4K for sanity:
-    nBlockMaxWeight = std::max<size_t>(4000, std::min<size_t>(MAX_BLOCK_WEIGHT - 4000, options.nBlockMaxWeight));
+    nBlockMaxWeight = std::clamp<size_t>(options.nBlockMaxWeight, 4000, MAX_BLOCK_WEIGHT - 4000);
```
Can we please add a static assert that `MAX_BLOCK_WEIGHT <= 8000`?

MarcoFalke replied:
Wouldn't it be better to have a clamp function that does this by default, internally and at compile time? The condition must hold for all call sites.

laanwj replied:
Though in most cases it's not possible to verify it at compile time, this would handle only the 'between two constexpr' case. In other cases there might be non-trivial error handling.

MarcoFalke replied:
For the cases where it is not possible to verify at compile time, crashing the node seems preferable over UB. Or is this something that sanitizers can spot?

laanwj commented:
Code review ACK 512dcf7d5b7ff815ce4fec2d6fe27cbb13aa3e33
I checked all the changes and they look correct to me.

MarcoFalke commented:
Slightly tend toward NACK. It is non-trivial to review (https://github.com/bitcoin/bitcoin/pull/23095#issuecomment-927779377) and easily introduces UB, if the second or third argument are not compile time constants (https://github.com/bitcoin/bitcoin/pull/23095#issuecomment-927780250).

hebasto commented:
> It is non-trivial to review ([#23095 (comment)](https://github.com/bitcoin/bitcoin/pull/23095#issuecomment-927779377)) and easily introduces UB.

Closing.

MarcoFalke commented:
Maybe write a `Clamp<int64_t, /*min=*/-1, /*max=*/3>(var);` wrapper that has static asserts built in, and another one `Clamp` that copies `std::clamp`'s signature that asserts/throws/returns optional on UB?
